raster/src/main/scala/geotrellis/raster/io/geotiff/OverviewStrategy.scala

```diff
@@ -16,20 +16,122 @@
+    * If the input list is not sorted, this function returns idx 0.
+    *
+    * Unless a particular strategy suggests otherwise, this method will clamp the returned
+    * index to the range of overviewCS.
+    * @param overviewCS a sorted list of resolutions, otherwise it is not guaranteed that it would work with all the input strategies.
+    * @param desiredCS  cellSize that would be searched
+    * @param strategy   overview strategy that would be used to search for the best matching CellSize
+    * @return           index of the closest cellSize. Each strategy behaves a bit different:
+    *                   Level: selects the passed index and if it is OOB clamps the index.
+    *                   Auto(n): selects the best matching cellSize (approximately) and adds n, requires the input resolutions list to be sorted.
+    *                   Base: always returns the best matching cellSize from the given list of CellSizes.
+    *                   AutoHigherResolution: selects the best matching cellSize (approximately) always selects a higher resolution,
+    *                   requires the input resolutions list to be sorted.
+    */
+  def selectOverview(
+    overviewCS: List[CellSize],
+    desiredCS: CellSize,
+    strategy: OverviewStrategy
+  ): Int = {
+    val maybeIndex = strategy match {
+      case Level(overviewIdx) => overviewIdx
+      case Auto(n)            => selectIndexByProximity(overviewCS, desiredCS, 0.5) + n
+      // returns the best resolution always
+      case Base               => overviewCS.indexOf(overviewCS.min)
+      // Returns -1 in case nothing was found or the input list was not sorted.
+      case AutoHigherResolution =>
+        val idx = selectIndexByProximity(overviewCS, desiredCS, 1.0)
+        // AutoHigherResolution defaults to the closest level
+        // idx > overviewCS.size means, that idx = overviewCS.size (last idx + 1)
+        if (idx > overviewCS.size) idx - 1 else idx
+    }
+    // clamp the index to fit the input list size
+    if (maybeIndex < 0) 0
+    else if (maybeIndex >= overviewCS.size) overviewCS.size - 1
+    else maybeIndex
+  }
+  
+  /**
+    * This method is unsafe. It's up to the selectOverview method to handle each index OOB error.
+    * @param overviewCS        a sorted list of resolutions, otherwise it is not guaranteed that it would work with all the input strategies.
+    * @param desiredCS          cellSize that would be searched  
+    * @param proximityThreshold threshold to defined the search proximity
+    * @return                   index of the closest cellSize. Returns -1 in case nothing was found or the input list was not sorted.  
+    */
+  private def selectIndexByProximity(overviewCS: List[CellSize], desiredCS: CellSize, proximityThreshold: Double): Int =
+    overviewCS.zipWithIndex.foldLeftM(Option.empty[(CellSize, Int)]) { case (acc, r @ (rightCZ, _)) =>
```

Comment:
I am not sure that with folding it is much easier to read, however it is more explicit about sorting.